File: Majority Element.py
class Solution:
    def majorityElement(self, nums):
        def recursiveMajority(nums, candidate, count, index):

            if index == len(nums):
                return candidate


            if count == 0:
                candidate = nums[index]


            count += (1 if nums[index] == candidate else -1)


            return recursiveMajority(nums, candidate, count, index + 1)


        return recursiveMajority(nums, None, 0, 0)

File: Majority Element.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var majorityElement = function(nums) {
    return recursiveMajority(nums, 0, null, 0);
};

function recursiveMajority(nums, index, candidate, count) {
    if (index === nums.length) {
        return candidate;
    }

    if (count === 0) {
        candidate = nums[index];
    }

    count += (nums[index] === candidate ? 1 : -1);
    return recursiveMajority(nums, index + 1, candidate, count);
}
